### PR TITLE
Added functions for deleting locked files.

### PR DESCRIPTION
**Pull request by heetbeet**
- Especially helpful for Excel locks.
- Want to use this in the uninstaller when Python is available to the installed program.

### CODE REVIEW

File: includes/rm_and_kill.py [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+    Like os.walk but generating actual Path objects and in absolute() form
+    """
+
+    path = Path(path).absolute()
+    if path.is_file():
+        yield path
+
+    else:
+        for i in _Path(path).walk():
+            yield Path(i).absolute()
```
Comment (rudolfbyker) [inline includes/rm_and_kill.py:170]: I think you mean `Path(i).resolve()` ? https://docs.python.org/3/library/pathlib.html#pathlib.Path.resolve

File: includes/rm_and_kill.py [hunk 2]
```diff
@@ -0,0 +1,253 @@
+from typing import Dict, List, Union, Iterator, Tuple
+from pathlib import Path
+from path import Path as _Path
```
Comment (rudolfbyker) [inline includes/rm_and_kill.py:3]: This package sucks. Change my mind. ;)

File: includes/rm_and_kill.py [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+import os
+import sys
+import subprocess
+import dataclasses
+import textwrap
+import tempfile
+from contextlib import contextmanager
+
+this_dir = Path(globals().get("__file__", r".\_")).absolute().parent
+handlebin = this_dir.joinpath("handle64.exe")
```
Comment (rudolfbyker) [inline includes/rm_and_kill.py:17]: We need some documentation of what this exe is and where it comes from.

File: includes/rm_and_kill.py [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+        try:
+            os.kill(pid, signal.SIGTERM)
+        except OSError as e:
+            pass
+
+        i += 1
+
+@contextmanager
+def handler():
+    # Put here what would ordinarily go in the `__enter__` method
```
Comment (rudolfbyker) [inline includes/rm_and_kill.py:74]: Don't document how the language works (if you really need to you can link to python docs). Rather document what you are trying to achieve / what the purpose of this function is. Also, use `"""` comments for function docs.

File: includes/rm_and_kill.py [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+    # In this case, there's nothing to do
+    try:
+        yield None# You can return something if you want, that gets picked up in the 'as'
+    except Exception as e:
+        print("no not possible")
```
Comment (rudolfbyker) [inline includes/rm_and_kill.py:79]: Do we really want to print this in production code??

File: includes/rm_and_kill.py [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+        p.kill()
+        if not fpath.exists():
+            return
+
+        while True:
+            try:
+                os.remove(fpath)
+                break
+            except PermissionError:
+                pass
```
Comment (rudolfbyker) [inline includes/rm_and_kill.py:128]: This will cause an busy loop (high CPU usage) when the file's permissions have changed for some reason. Not good.

**Review by rudolfbyker — CHANGES_REQUESTED** [review]

File: includes/rm_and_kill.py (hunk 2 above)
Comment (heetbeet) [reply includes/rm_and_kill.py:3]: - It has the .walk() generator that returns full _Path objects (and not _dir, files, dirs like os.walk())
- It has context manager for changing directories build in.

File: includes/rm_and_kill.py (hunk 3 above)
Comment (heetbeet) [reply includes/rm_and_kill.py:17]: Will add:
https://docs.microsoft.com/en-us/sysinternals/downloads/handle

File: includes/rm_and_kill.py (hunk 5 above)
Comment (heetbeet) [reply includes/rm_and_kill.py:79]: Oops, some leftover copy/pasted code

File: includes/rm_and_kill.py (hunk 6 above)
Comment (heetbeet) [reply includes/rm_and_kill.py:128]: Okay, I only use this in tests, but that is no excuse for a sloopy function. Do you think I should add a time.sleep or what is the best way to overcome a busy loop?

File: includes/rm_and_kill.py (hunk 1 above)
Comment (heetbeet) [reply includes/rm_and_kill.py:170]: Hmm, `.absolute()` works just fine. Wonder why it isn't documented. Will change to `.resolve()` though.

File: includes/rm_and_kill.py (hunk 2 above)
Comment (rudolfbyker) [reply includes/rm_and_kill.py:3]: I will investigate again. Anyway, it's lacking a few things that I use often in the standard pathlib, which kind of sucks. But if we need it for context manager and walk, that makes sense.

File: includes/rm_and_kill.py (hunk 6 above)
Comment (rudolfbyker) [reply includes/rm_and_kill.py:128]: If it's only used in tests, but it in a separate file and write in the doc string "this is only used in tests".

Never mind busy loop, why do you need a loop at all? Just let the test fail if you fail to remove the file. Or if you really need to wait for the process to end, use p.kill() followed by p.join() or something like that. Worst case, if you really need to loop, add time.sleep() and a maximum number of retries before throwing an error.

File: includes/rm_and_kill.py (hunk 1 above)
Comment (heetbeet) [reply includes/rm_and_kill.py:170]: Hmm, if I change all my `.absolute()` to `.resolve()1 some of my tests starts failing. It seems like Python doesn't like to `.resolve()` a filepath if some permissions are set on the file. I will change `.resolve()` over here, and at a few other places though.

Comment (heetbeet) [reply includes/rm_and_kill.py:170]: Oh and wait, I don't want to `resolve` symlinks when walking.
There is a whole discussion on this:
https://discuss.python.org/t/pathlib-absolute-vs-resolve/2573/

This answer sums it up nicely:
https://discuss.python.org/t/pathlib-absolute-vs-resolve/2573/5

Damn, this is annoying. Maybe I should just add my own abspath.

Comment (heetbeet) [reply includes/rm_and_kill.py:170]: -1 for pathlib!

Comment (heetbeet) [reply includes/rm_and_kill.py:170]: I fixed it by switching to `Path(os.path.abspath(path))` wherever I'm unsure if `.resolve()` might give some crazy error.

File: includes/rm_and_kill.py (hunk 6 above)
Comment (heetbeet) [reply includes/rm_and_kill.py:121]: Okay, I added that this is used for testing.

**Pull request merged**